src/main/java/io/strimzi/kproxy/internal/KafkaProxyBackendHandler.java

```diff
@@ -75,6 +75,10 @@ public void channelRead(final ChannelHandlerContext ctx, Object msg) {
             inboundChannel.writeAndFlush(msg, inboundCtx.voidPromise());
             unflushedWrites = false;
         }
+
+        if (msg instanceof BufferHolder) {
+            ((BufferHolder) msg).release();
```

Comment:
I don't know if this is yet the very last one to use the msg,,,
The kafka builder is creating duplicates of the held buffers (meaning that they ideally share the same content but not the NIO ByteBuffer instance on it), hence to save having weird errors to happen (due to unwanted shared buffers), we should release it just when the msg instance won't be used anymore.
This will add the additional requirement of tracking the msg lifecycle/accsses by using a boolean `released` field on the `BufferHolder` itself. This will save us any misuses.

Reply:
> I don't know if this is yet the very last one to use the msg

I think it is the last one. It is the [handler of the outbound channel](https://github.com/kroxylicous/kproxy/blob/d539cdfb8419b345dbf095021c9af3568cb5e932/src/main/java/io/strimzi/kproxy/internal/KafkaProxyFrontendHandler.java#L94) and I see in the debugger that this method is invoked after the last handler in the chain (the "backend-application" logging handler), so I think everything is alright.

> we should release it just when the msg instance won't be used anymore.

This is what's happening from what I can say. `release()` is only called after the response on the outbound channel has been encoded again. Let's have a quick call if you still believe there's something wrong?

Reply:
I now remember why I was proposing you to make the `DecodedFrame` to extend the `ReferenceCounted` interface, see
```java
                try {
                    this.encode(ctx, cast, buf);
                } finally {
                    ReferenceCountUtil.release(msg);
                }
```
This is what happen within `MessageToByteEncoder` (parent of `KafkaMessageEncoder`): in short, you can decide to use a custom known type of Netty and the infrastructure would provide proper releasing of `msg`

Reply:
Ok, gotcha. For my understanding: is what's currently proposed here functionally correct (just less Netty-idiomatic) and we could go with it for now and cycle back to making this `ReferenceCounted` later on, or is there any actual problem with the change in this PR in its current form?

Reply:
For the sake of closure: we are using `ReferenceCounted` now.